cmd/kube-proxy/app/server.go

```diff
@@ -90,123 +77,55 @@ const (
-	fs.DurationVar(
-		&o.config.Conntrack.TCPCloseWaitTimeout.Duration, "conntrack-tcp-timeout-close-wait",
-		o.config.Conntrack.TCPCloseWaitTimeout.Duration,
-		"NAT timeout for TCP connections in the CLOSE_WAIT state")
-	fs.BoolVar(&o.config.EnableProfiling, "profiling", o.config.EnableProfiling, "If true enables profiling via web interface on /debug/pprof handler.")
-	fs.StringVar(&o.config.IPVS.Scheduler, "ipvs-scheduler", o.config.IPVS.Scheduler, "The ipvs scheduler type when proxy mode is ipvs")
-	fs.StringSliceVar(&o.config.NodePortAddresses, "nodeport-addresses", o.config.NodePortAddresses,
-		"A string slice of values which specify the addresses to use for NodePorts. Values may be valid IP blocks (e.g. 1.2.3.0/24, 1.2.3.4/32). The default empty string slice ([]) means to use all local addresses.")
-	fs.Var(flag.NewMapStringBool(&o.config.FeatureGates), "feature-gates", "A set of key=value pairs that describe feature gates for alpha/experimental features. "+
-		"Options are:\n"+strings.Join(utilfeature.DefaultFeatureGate.KnownFeatures(), "\n"))
-}
+	cmd := &cobra.Command{
+		Use: "kube-proxy",
+		Long: `The Kubernetes network proxy runs on each node. This
+reflects services as defined in the Kubernetes API on each node and can do simple
+TCP and UDP stream forwarding or round robin TCP and UDP forwarding across a set of backends.
+Service cluster IPs and ports are currently found through Docker-links-compatible
+environment variables specifying ports opened by the service proxy. There is an optional
+addon that provides cluster DNS for these cluster IPs. The user must create a service
+with the apiserver API to configure the proxy.`,
+		Run: func(cmd *cobra.Command, args []string) {
+			verflag.PrintAndExitIfRequested()
+			utilflag.PrintFlags(cmd.Flags())
 
-func NewOptions() *Options {
-	return &Options{
-		config:      new(kubeproxyconfig.KubeProxyConfiguration),
-		healthzPort: ports.ProxyHealthzPort,
-		scheme:      scheme.Scheme,
-		codecs:      scheme.Codecs,
-		CleanupIPVS: true,
+			if err := initForOS(opts.WindowsService); err != nil {
+				glog.Fatalf("failed OS init: %v", err)
+			}
+
+			cmdutil.CheckErr(Complete(opts))
+			cmdutil.CheckErr(Validate(opts, args))
+			cmdutil.CheckErr(Run(opts))
+		},
 	}
+
+	opts.AddFlags(cmd.Flags())
+
+	cmd.MarkFlagFilename("config", "yaml", "yml", "json")
+
+	return cmd
 }
 
 // Complete completes all the required options.
-func (o *Options) Complete() error {
+func Complete(o *options.Options) error {
 	if len(o.ConfigFile) == 0 && len(o.WriteConfigTo) == 0 {
 		glog.Warning("WARNING: all flags other than --config, --write-config-to, and --cleanup are deprecated. Please begin using a config file ASAP.")
-		o.applyDeprecatedHealthzPortToConfig()
+		o.ApplyDeprecatedHealthzPortToConfig()
 	}
 
 	// Load the config file here in Complete, so that Validate validates the fully-resolved config.
 	if len(o.ConfigFile) > 0 {
```

Comment:
FYI there was a regression that @resouer identified to me on slack: with my original PRs, if you happened to run `kube-proxy --config foo --write-config-to bar`, it would ignore `--config foo` and write the default config values to `bar`. Before https://github.com/kubernetes/kubernetes/pull/55261, we set up a default config struct, validated those defaults, and finally loaded the config from the file, meaning that the actual config wasn't being validated. With that PR, we now load the config prior to validation.

This has the side effect that when you run `kube-proxy --config foo --write-config-to bar`, it first merges any configuration from `foo` with default values for any unset fields, then it writes to disk.

I think we should make `--config` and `--write-config-to` mutually exclusive to avoid this situation. WDYT?